glusterd2/pmap/registry.go

```diff
@@ -0,0 +1,207 @@
+package pmap
+
+import (
+	"expvar"
+	"fmt"
+	"net"
+	"sync"
+)
+
+const (
+	gfIanaPrivPortsStart = 49152
+	gfPortMax            = 65535
+)
+
+// PortState represents the type of state of the port
+type PortState int32
+
+// List of states an individual port can be in
+const (
+	GfPmapPortFree PortState = iota
+	GfPmapPortForeign
+	GfPmapPortLeased
+	GfPmapPortInUse
+)
+
+type portStatus struct {
+	Port  int       `json:"port"`
+	State PortState `json:"state"`
+	Conn  net.Conn  `json:"-"`
+}
+
+type brickStatus struct {
+	Port     int  `json:"port"`
+	SignedIn bool `json:"signed-in"`
+}
+
+type registryType struct {
+	sync.RWMutex
+	basePort  int
+	lastAlloc int
+	ports     [gfPortMax + 1]portStatus
+	bricks    map[string]*brickStatus
+
+	portLockFds map[int]int
+}
+
+func (r *registryType) init() {
+	for i := r.basePort; i <= gfPortMax; i++ {
+		// TODO: There is a race in this check when there are multiple
+		// glusterd2 instances running on the same machine.
+		if isPortFree(i) {
+			r.ports[i].State = GfPmapPortFree
+		} else {
+			r.ports[i].State = GfPmapPortForeign
+		}
+		r.ports[i].Port = i
+	}
+}
+
+// Allocate finds a free port and returns the port number. The
+// allocated free port will be in leased state until the brick
+// process it was leased to sends a SIGN IN request to glusterd2.
+func (r *registryType) Allocate(brickpath string) int {
+	r.Lock()
+	defer r.Unlock()
+
+	var port int
+	for p := r.basePort; p <= gfPortMax; p++ {
+		if r.ports[p].State == GfPmapPortFree || r.ports[p].State == GfPmapPortForeign {
+			// check if a port is free holding a lock on the port's "file"
+			if r.TryPortLock(p) {
+				if isPortFree(p) {
+					r.ports[p].State = GfPmapPortLeased
+					r.bricks[brickpath] = &brickStatus{
+						Port:     p,
+						SignedIn: false,
+					}
+					port = p
+					// keep port file locked if port was
+					// found to be free and we leased it to
+					// a brick
+					break
+				} else {
+					r.ports[p].State = GfPmapPortForeign
+					r.PortUnlock(p)
+				}
+			}
+		}
+	}
+
+	if port == 0 {
+		panic("registryAlloc() we ran out of free ports!")
```

Comment:
Don't panic. Return an error so that the brick doesn't start and a proper error is returned to the user.

Reply:
I just remembered why I marked this as `panic`. The call flow upwards reaches daemon interface's `Args() []string` method which doesn't return error. An alternative is to assign port right when `brick.Glusterfsd` object gets created, which isn't clean.

To avoid making invasive changes outside of pmap package, I'll leave it as is for now. When glusterfs change https://review.gluster.org/14670 gets in, this port allocation code will be disposed off anyway.

Reply:
Done but the error is not propagated.